### ISSUE TITLE
Add support for Databricks Runtime 7.0 (Scala 2.12, Spark 3.0)

### ISSUE THREAD

**Issue by al-lac** [mspnp/spark-monitoring#92]
Please add support for the new Databricks Runtime 7.0 which is based on Scala 2.12 and Spark 3.0. 

Adding a maven profile for it to **pom.xml** does not work because of too many changes in the new Spark and Scala versions:
```xml
<profile>
            <id>scala-2.12_spark-3.0.0</id>
            <activation>
                <activeByDefault>true</activeByDefault>
            </activation>
            <properties>
                <codahale.metrics.version>3.1.5</codahale.metrics.version>
                <commons-codec.version>1.10</commons-codec.version>
                <commons.httpclient.version>4.5.6</commons.httpclient.version>
                <fasterxml.jackson.databind.version>2.10.0</fasterxml.jackson.databind.version>
                <jetty.version>9.4.18.v20190429</jetty.version>
                <log4j.version>1.2.17</log4j.version>
                <slf4j.version>1.7.30</slf4j.version>
                <spark.version>3.0.0</spark.version>
                <scala.version>2.12.10</scala.version>
                <scala.compat.version>2.12</scala.compat.version>
            </properties>
        </profile>
```

**Comment by snelluri** [comment]
I am also looking for this. @al-lac, By any chance were you able to get this resolved?

**Comment by al-lac** [comment]
Unfortunately, no, not yet. I tried to update the code myself to support the new Scala and Spark versions. But there are too many changes needed for my limited knowledge of Scala and Spark. 

Maybe some of my contacts at microsoft can help push this further.

**Comment by snelluri** [comment]
Thanks @al-lac  for your quick response and good luck with your plans.

**Comment by GBBBAS** [comment]
Also need this as we are moving to Spark 3.0 and DBR 7.2(LTS). An update for this version would be greatly appreciated

**Comment by bmoregeo** [comment]
Also chiming in about adding support for DB 7+. Do we have a rough timeline? My understanding is 6.4 will no longer be supported in April 2021. Thanks!

**Comment by al-lac** [comment]
Hey everyone, the response from microsoft was that there are no plans to update the repository to support Databricks Runtime 7+ this quarter. 

I tried my best to get it to run with my limited scala knowledge and created a branch in my fork [here](https://github.com/al-lac/spark-monitoring/tree/add-db-runtime-7-support). The branch only supports Scala 2.12 + Spark 3.0. I had to disable some tests to get the build to run through. 
Logging with the new .jars worked on my Databricks Runtime 7.0 Clusters.

**Comment by timo-aho** [comment]
I got some information from Microsoft side. Apparently there is progress on this.

**Comment by Emresav** [comment]
I also need Spark 3.0.0 DB Runtime 7.x support as soon as possible 
@hallihan - Do we have it planned to be supported?
Thanks for your help all!

**Comment by GBBBAS** [comment]
Not sure if this would a be a workaround that could be useful for some people waiting for this update? https://docs.microsoft.com/en-us/azure/databricks/administration-guide/account-settings/azure-diagnostic-logs#:~:text=Configure%20diagnostic%20log%20delivery,-Note&text=Log%20in%20to%20the%20Azure,Click%20Turn%20on%20diagnostics.&text=Enter%20a%20name%20for%20the%20logs%20to%20create.

**Comment by hallihan** [comment]
We have work planned on this for later this quarter, I'm afraid I can't provide an ETA at this point.

**Comment by Nemeczek** [comment]
> Not sure if this would a be a workaround that could be useful for some people waiting for this update? https://docs.microsoft.com/en-us/azure/databricks/administration-guide/account-settings/azure-diagnostic-logs#:~:text=Configure%20diagnostic%20log%20delivery,-Note&text=Log%20in%20to%20the%20Azure,Click%20Turn%20on%20diagnostics.&text=Enter%20a%20name%20for%20the%20logs%20to%20create.

Well, it does not because these logs are available in the next 24 hours. So for analysis purposes, it can do but for any live monitoring, it won't. It is a bit bizarre that something that urgent is not prioritised at all.

**Comment by al-lac** [comment]
> Well, it does not because these logs are available in the next 24 hours. So for analysis purposes, it can do but for any live monitoring, it won't. It is a bit bizarre that something that urgent is not prioritised at all.

Hi @Nemeczek, the logs should now be available with little to no delay.

**Comment by Nemeczek** [comment]
> > Well, it does not because these logs are available in the next 24 hours. So for analysis purposes, it can do but for any live monitoring, it won't. It is a bit bizarre that something that urgent is not prioritised at all.
> 
> Hi @Nemeczek, the logs should now be available with little to no delay.

They had to change it because I still remember this user voice:
https://feedback.azure.com/forums/909463-azure-databricks/suggestions/39417028-azure-diagnosticks-logs-are-collected-with-up-to-2

And now docs are stating that "On any given day, Azure Databricks delivers at least 99% of diagnostic logs within the first 15 minutes". Which is not ideal but somehow useful.

**Comment by Emresav** [comment]
> > Not sure if this would a be a workaround that could be useful for some people waiting for this update? https://docs.microsoft.com/en-us/azure/databricks/administration-guide/account-settings/azure-diagnostic-logs#:~:text=Configure%20diagnostic%20log%20delivery,-Note&text=Log%20in%20to%20the%20Azure,Click%20Turn%20on%20diagnostics.&text=Enter%20a%20name%20for%20the%20logs%20to%20create.
> 
> Well, it does not because these logs are available in the next 24 hours. So for analysis purposes, it can do but for any live monitoring, it won't. It is a bit bizarre that something that urgent is not prioritised at all.



Will this send all the Databricks logs to the target location? including Spark app logs, executor and driver logs etc.

**Comment by Nemeczek** [comment]
> > > Not sure if this would a be a workaround that could be useful for some people waiting for this update? https://docs.microsoft.com/en-us/azure/databricks/administration-guide/account-settings/azure-diagnostic-logs#:~:text=Configure%20diagnostic%20log%20delivery,-Note&text=Log%20in%20to%20the%20Azure,Click%20Turn%20on%20diagnostics.&text=Enter%20a%20name%20for%20the%20logs%20to%20create.
> > 
> > 
> > Well, it does not because these logs are available in the next 24 hours. So for analysis purposes, it can do but for any live monitoring, it won't. It is a bit bizarre that something that urgent is not prioritised at all.
> 
> Will this send all the Databricks logs to the target location? including Spark app logs, executor and driver logs etc.

No it does not. 
It only delivers this:
![image](https://user-images.githubusercontent.com/1111404/96436634-4f089b80-1206-11eb-980c-995321eb0c51.png)
Which is all related to the Databricks but not to the internal Spark logs. To get the complete picture you need this library.

**Comment by marcin-vt** [comment]
Would be nice to have this, a specially when taking into account that runtimes based on spark 2.* will run out of support early next year, 

THNX!

**Comment by dfanesiDB** [comment]
Hello Everyone,

I just did the implementation for spark 3 and now I am testing it. I was able to maintain the same tests and the same codebase but some methods were expecting more parameters on spark 3, so I used reflections in the tests to make it working with both versions. Not the nicest solution but it works.

I am planning to do a pull request soon.

**Comment by dfanesiDB** [comment]
Hello Everyone, 

I have added the support for Spark 3 and made the same tests work with both versions. I am now in the process of manually testing the jars on the DBRs.

I will open a pull request as soon as I completed the QA.

**Comment by hallihan** [comment]
@dfanesiDB that's great!  I should have time to review the pull request next week.

**Comment by esdhrgl1984** [comment]
Hi All,
Our current project is using DBR 6.2 with scala 2.11 and spark 2.4.4. We are planning to move to later version of DBR. Can someone confirm if spark-monitoring with work in our case or this is a hard dependency?

**Comment by al-lac** [comment]
Hi @esdhrgl1984, once #115 is merged you can use DBR with version 7.

**Comment by esdhrgl1984** [comment]
Hi @al-lac , Apologies if my question was vague. We are looking to make this work on DBR 6.2/Spark 2.4.4 till we migrate to other DBR/Spark version. 
Based on the documentations, currently the spark-monitoring library is restricted to specific DBR/spark/scala versions. Correct me if this is not correct? Also, README could list these dependencies more clearly.

**Comment by hallihan** [comment]
@esdhrgl1984 the intent is to update the library to support all current supported releases from: <https://docs.microsoft.com/en-us/azure/databricks/release-notes/runtime/releases#--supported-releases>

6.2 is out of support as of Jun 03, 2020 (<https://docs.microsoft.com/en-us/azure/databricks/release-notes/runtime/releases#--end-of-support-history>).

With that said, the library has been built and worked with older versions previously.  If you look at the profiles in [build.sh](build.sh) and [src/pom.xml](src/pom.xml) you should be able to figure out how to build the library with your versions.

**Comment by mcheedle** [comment]
This is terrific! Thank you, I can't wait to deploy.

**Comment by hallihan** [comment]
Big thanks to @dfanesiDB for the contribution!
